crates/miden-objects/src/note/note_tag.rs

```diff
@@ -248,23 +276,54 @@ impl fmt::Display for NoteTag {
 
 impl From<u32> for NoteTag {
     fn from(value: u32) -> Self {
-        Self(value)
+        let prefix = value >> 30;
+        let lower_30_bits = value & 0x3fffffff; // Mask to get the lower 30 bits
```

Comment:
We shouldn't need this at all if we include the prefix in the values. This function currently masks out the prefix but we should include it. If we remove this, the logic here should become much easier.